rfcs/text/0000-query-by-metadata-schema-versions.md

```diff
@@ -0,0 +1,202 @@
+tests are setup, subscriptions will be modified to remove any restrictions on metadata schema versions in order to
+eagerly surface incompatibilities.
+
+
+### User Stories
+
+*Share the [User Stories](https://www.mountaingoatsoftware.com/agile/user-stories) motivating this RFC.*
+* As a data wrangle, I would like to be able to push new data adhering to the latest metadata schema without having to
+wait for every single component to be compatible with the new schema versions.
+* As a DCP developer working on a downstream component (e.g. Matrix Service, Secondary Analysis, etc.) I would like to
+receive notification about new data adhering to new metadata schema versions as soon as I am ready without having to
+wait for other DCP components to also be compatible. 
+
+## Detailed Design
+
+There are two questions fueling the design of this system:
+
+1) How will the schema version be noted in the metadata schema?
+2) Will ElasticSearch, JMESPath, and Query Service all be able to issue range queries against the format of the schema
+version?
+
+Today, the version of the metadata schema is embedded in the `described_by` field which contains the schema version
+embedded in the URL; an example value is “https://schema.humancellatlas.org/type/biomaterial/15.3.0/donor_organism” and
+the schema version is 15.3.0. Unfortunately embedding the version within the URL does not facilitate searching for a
+range of data between schema versions. 
+
+### New schema version fields in metadata schema
+We propose adding two new required fields to each list of schema properties:
+
+```
+"schema_major_version": {
+      "description": "The major version number of the schema.",
+      "type": "integer",
+      "pattern": "^[1-9]\\d*$",
+      "example": "4"
+},
+"schema_minor_version": {
+      "description": "The minor version number of the schema.",
+      "type": "integer",
+      "pattern": "^\\d+$",
+      "example": "6"
+}
+```
+
+We have noticeably left off adding the patch version; this is due to the fact that patch version increments only affect
+documentation about the metadata and therefore can have no effect on any code that uses the metadata schema. Since it
+does not affect forward or backwards compatibility we have chosen not to represent it explicitly in the schema which
+means that subscribing to data based on the patch version is not supported.
+
+We have decided not to represent the full version as `x.y.z` in the schema_version because formatting in this way is not
+compatible with the way that ElasticSearch performs range queries. Range queries in ElasticSearch can only be done on
+fields that are either an integer or a floating point number (or a datetime).
+
+### ElasticSearch Query for Metadata Schema Version
+```
+# Query for files where the metadata schema version of the project_json metadata schema # is greater than or equal to
+# 1.0 but less than 3.2
```

Comment:
Also, I'm assuming downstream components defining their subscription ES queries will potentially have to specify a similar `should` clause to the above for every metadata file we want to control. While I doubt this will affect performance, I will note that queries will become relatively long/verbose. Granted, this sounds like a short term solution before integration tests are implemented.

Reply:
Woah! Thank you for the query @calvinnhieu !!

Yes, there is ongoing work to deprecate ES queries so this should be a short-term solution.

Reply:
Fixed both the ES and JMESPath queries. Thank you so much!